### PR TITLE
Correctly cast `fill_value` in constant initializer

### PR DESCRIPTION
**Pull request by hvy**
An alternative to https://github.com/chainer/chainer/pull/8043 on request per @toslunar.

### CODE REVIEW

File: chainer/initializers/constant.py [hunk 1]
```diff
@@ -51,13 +51,16 @@ def __call__(self, array):
...
 
+        fill_value = self.fill_value
         # Calling copy to ensures that the fill_value array
         # is moved to the device where array resides
-        if isinstance(self.fill_value, chainer.get_array_types()):
-            backend.copyto(array, self.fill_value)
+        if isinstance(fill_value, chainer.get_array_types()):
+            if self.dtype is not None:
+                fill_value = fill_value.astype(self.dtype, copy=False)
+            backend.copyto(array, fill_value)
```
Comment (niboshi) [inline chainer/initializers/constant.py]: I think these 3 lines can be written simply as:
```py
backend.copyto(array, self.fill_value.astype(array.dtype, copy=False))
```

Comment (hvy) [reply chainer/initializers/constant.py]: It fails for ChainerX. Should I add a TODO comment?

Comment (niboshi) [reply chainer/initializers/constant.py]: How does it fail?

Comment (hvy) [reply chainer/initializers/constant.py]:
```python
chainerx.arange(1).astype(dtype=None) # TypeError: Dtype not understood: None
numpy.arange(1).astype(dtype=None) # array([0.]) with float64 dtype
```

Comment (hvy) [reply chainer/initializers/constant.py]: Sorry now I noticed you wrote `array.dtype`.

Comment (hvy) [reply chainer/initializers/constant.py]: Amended changes, PTAL.

**Comment by niboshi** [comment]
LGTM
@toslunar Could you review?

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
This bug was introduced in #8022: no need to backport.

**Pull request merged**